File: 2685-count-the-number-of-complete-components/2685-count-the-number-of-complete-components.py
class Solution:
    def countCompleteComponents(self, n: int, edges: list[list[int]]) -> int:
        res = 0
        graph = [[] for _ in range(n)]
        nodes = set()
        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)
        def bfs(node: int) -> List[int]:
            q = [node]
            visited = {node}
            component = [node]
            idx = 0
            while idx < len(q):
                cur = q[idx]
                idx += 1
                for neighbor in graph[cur]:
                    if neighbor not in visited:
                        visited.add(neighbor)
                        component.append(neighbor)
                        q.append(neighbor)
            return component
        for i in range(n):
            if i not in nodes:
                component = bfs(i)
                connected = True
                for node in component:
                    if len(graph[node]) != len(component) - 1:
                        connected = False
                        break
                if connected:
                    res += 1
                for node in component:
                    nodes.add(node)
        return res

File: 2685-count-the-number-of-complete-components/2685-count-the-number-of-complete-components.cpp
class Solution {
public:
    int countCompleteComponents(int n, vector<vector<int>>& edges) {
        int res = 0;
        vector<vector<int>> graph(n);
        unordered_set<int> nodes;
        for(int i = 0; i < edges.size(); i++)
        {
            graph[edges[i][0]].push_back(edges[i][1]);
            graph[edges[i][1]].push_back(edges[i][0]);
        }
        for(int i = 0; i < n; i++)
        {
            if(!nodes.count(i))
            {
                vector<int> component = {};
                nodes.insert(i);
                component.push_back(i);
                bfs(graph, nodes, i, component);
                bool connected = true;
                for(int j = 0; j < component.size(); j++)
                {
                    if(graph[component[j]].size() != component.size() - 1)
                    {
                        connected = false;
                        break;
                    }
                }
                if(connected)
                {
                    res++;
                }
            }
        }
        return res;
    }
private:
    void bfs(vector<vector<int>>& graph, unordered_set<int>& nodes, int node, vector<int>& component)
    {
        queue<int> q;
        q.push(node);
        while(!q.empty())
        {
            int cur = q.front();
            q.pop();
            for(int i = 0; i < graph[cur].size(); i++)
            {
                if(!nodes.count(graph[cur][i]))
                {
                    q.push(graph[cur][i]);
                    component.push_back(graph[cur][i]);
                    nodes.insert(graph[cur][i]);
                }
            }
        }
    }
};